### PR TITLE
Add support for dual mono setups (AVPro)

### PR DESCRIPTION
**Pull request by fundale**
Optional support for dual mono audio

### CODE REVIEW

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs [hunk 1]
```diff
@@ -765,15 +767,21 @@ public void SendAudioOutputData()
                 }
                 else
                 {
-                    audioSource.GetOutputData(_audioFramesR, 1);
+                    if (VRC.SDKBase.Utilities.IsValid(optionalRightAudioSource))
```
Comment (float3) [inline Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs:770]: issue (blocking): any calls to VRC namespace stuff should be wrapped with `#if UDONSHARP` or `#if VRC_SDK_VRCSDK3` depending on if that codepath should be taken in Avatar Projects, in this case I believe it should be `UDONSHARP`

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs [hunk 2]
```diff
@@ -765,15 +767,21 @@ public void SendAudioOutputData()
...
                 _rightChannelTestCounter--;
             }
             else
             {
-                _rightChannelTestCounter = _rightChannelTestDelay;      // reset test countdown
-                _audioFramesR[0] = 0f;                                  // reset tested array element to zero just in case
-                audioSource.GetOutputData(_audioFramesR, 1);            // right channel test
+                _rightChannelTestCounter = _rightChannelTestDelay;                  // reset test countdown
+                _audioFramesR[0] = 0f;                                              // reset tested array element to zero just in case
+                if (VRC.SDKBase.Utilities.IsValid(optionalRightAudioSource))        // check if dual mono is present
```
Comment (float3) [inline Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs:781]: issue (blocking): any calls to VRC namespace stuff should be wrapped with `#if UDONSHARP` or `#if VRC_SDK_VRCSDK3` depending on if that codepath should be taken in Avatar Projects, in this case I believe it should be `UDONSHARP`

**Review by float3 — CHANGES_REQUESTED** [review]

**Review by float3 — APPROVED** [review]
lgtm

**Comment by pema99** [comment]
The change looks fine to me, but could you enlighten me a bit on how this should be used? It looks like it will just always pull the data from the newly added audio source field if it isn't null. Isn't this for switching between 2 audio sources? How do you actually switch?

**Comment by fundale** [comment]
> @fundale The change looks fine to me, but could you enlighten me a bit on how this should be used? It looks like it will just always pull the data from the newly added audio source field if it isn't null. Isn't this for switching between 2 audio sources? How do you actually switch?

This isn't for switching between two separate videoplayers or audiosources,
it's intention is to provide a way for video players using separate left and right speakers to function better.
Like when using AVPro through ProTV, VideoTXL, USharpVideo, etc.

As some audio may be stereo encoded with temporary mono right audio, that currently would not be captured when setup with AVPro unless using Stereo Mix

**Comment by pema99** [comment]
So how exactly is a user intended to use this? What do they plug into the optional right source slot? The only difference I see in code is the channel they sample from, and only one is used at any given time

**Comment by fundale** [comment]
> So how exactly is a user intended to use this? What do they plug into the optional right source slot? The only difference I see in code is the channel they sample from, and only one is used at any given time

If they're using a 2D Stereo mix source nothing changes with how they use it,

But for users using AVPro with dual mono they would have the left audio source remain in the current Audio Source slot, with the right audio source going into the Optional Right Audio Source.

Some video players already automate connecting to AudioLink so they may be able to auto configure it accordingly for users

It should still sample from left and right at the same time and combine them the same way AudioLink already does

**Comment by pema99** [comment]
Oh, I was misreading the diff. It looks fine to me with your explanation.

**Review by pema99 — APPROVED** [review]

**Pull request merged**